File: examples/app-redux-android-todolist/src/debug/java/com/redux/StateConverter.java
package com.redux;

import com.google.common.reflect.TypeToken;
import com.google.gson.Gson;

import java.util.ArrayList;
import java.util.List;

class StateConverter implements Converter<TodoState> {
    private final Gson gson;

    StateConverter(Gson gson) {
        this.gson = gson;
    }

    @Override public TodoState fromJson(String json) {
        final List<Todo> todos = gson.fromJson(json, new TypeToken<ArrayList<Todo>>() {}.getType());
        return TodoState.from(todos);
    }

    @Override public String toJson(TodoState state) {
        return gson.toJson(state.getTodoList());
    }
}

File: examples/app-redux-android-todolist-kotlin/src/debug/kotlin/com/redux/StateConverter.kt
package com.redux

import com.github.salomonbrys.kotson.fromJson
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonIOException
import com.redux.devtools.Converter

class StateConverter(val gson: Gson) : Converter<TodoListState> {

    override fun fromJson(json: String): TodoListState {
        val list = gson.fromJson<List<Todo>>(json)
        return if (list != null) TodoListState(list, false) else throw JsonIOException("Could not parse $json")
    }

    override fun toJson(state: TodoListState): String {
        return gson.toJson(state.list)
    }
}
